### DISCUSSION TITLE
I cannot get pwgan to converge sound good

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#209] [Archived]
<USER>>>> georroussos<br/> <DATE> [May 23, 2020, 9:06am]<br/> <br/> Hi,<br/> <br/> I have been trying to finetune the LJSpeech PWGAN model to my single<br/> speaker, but it refuses to sound better. The voice has indeed changed,<br/> but it's gotten to a quality threshold where it sounds deeper, with some<br/> noise and isn't improving. I have been training for 2 days now -- the<br/> loss doesn't really seem to be improving, only spiking around. First I<br/> tried with a batch size of 4 and then I tried a combination of 2 and<br/> distributed training, to cut the time.<br/> <br/> Does anyone have any ideas?<br/> <br/> allow_cache: true<br/> format: 'hdf5'<br/> audio:<br/> clip_norm: true<br/> do_trim_silence: true<br/> frame_length_ms: 50<br/> frame_shift_ms: 12.5<br/> max_norm: 4<br/> hop_length: 275<br/> win_length: 1100<br/> mel_fmax: 8000.0<br/> mel_fmin: 0.0<br/> min_level_db: -100<br/> num_freq: 1025<br/> num_mels: 80<br/> preemphasis: 0.98<br/> ref_level_db: 20<br/> sample_rate: 22050<br/> signal_norm: true<br/> sound_norm: false<br/> symmetric_norm: true<br/> trim_db: 60<br/> augment: false<br/> batch_max_steps: 28875<br/> batch_size: 2<br/> discriminator_grad_norm: 1<br/> discriminator_optimizer_params:<br/> eps: 1.0e-06<br/> lr: 5.0e-05<br/> weight_decay: 0.0<br/> discriminator_params:<br/> dropout: 0.0<br/> gate_channels: 64<br/> in_channels: 1<br/> kernel_size: 3<br/> layers: 30<br/> nonlinear_activation: LeakyReLU<br/> nonlinear_activation_params:<br/> negative_slope: 0.2<br/> out_channels: 1<br/> residual_channels: 32<br/> skip_channels: 32<br/> stacks: 3<br/> use_weight_norm: true<br/> discriminator_scheduler_params:<br/> gamma: 0.5<br/> step_size: 200000<br/> discriminator_train_start_steps: 100000<br/> discriminator_type: ResidualParallelWaveGANDiscriminator<br/> distributed: false<br/> eval_interval_steps: 1000<br/> generator_grad_norm: 10<br/> generator_optimizer_params:<br/> eps: 1.0e-06<br/> lr: 0.001<br/> weight_decay: 0.0<br/> generator_params:<br/> aux_channels: 80<br/> aux_context_window: 2<br/> dropout: 0.0<br/> gate_channels: 128<br/> in_channels: 1<br/> kernel_size: 3<br/> layers: 30<br/> out_channels: 1<br/> residual_channels: 64<br/> skip_channels: 64<br/> stacks: 3<br/> upsample_net: ConvInUpsampleNetwork<br/> upsample_params:<br/> upsample_scales:<br/> - 5<br/> - 5<br/> - 11<br/> use_weight_norm: true<br/> generator_scheduler_params:<br/> gamma: 0.5<br/> step_size: 200000<br/> generator_type: ParallelWaveGANGenerator<br/> hop_size: 275<br/> lambda_adv: 4.0<br/> log_interval_steps: 100<br/> num_save_intermediate_results: 4<br/> num_workers: 2<br/> pin_memory: true<br/> rank: 0<br/> remove_short_samples: true<br/> save_interval_steps: 2000<br/> train_max_steps: 1000000<br/> verbose: 1<br/> <br/> ###########################################################<br/> # STFT LOSS SETTING #<br/> ###########################################################<br/> stft_loss_params:<br/> fft_sizes: [1024, 2048, 512] # List of FFT size for STFT-based loss.<br/> hop_sizes: [120, 240, 50] # List of hop size for STFT-based loss<br/> win_lengths: [600, 1200, 240] # List of window length for STFT-based loss.<br/> window: 'hann_window' # Window function for STFT-based loss<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/i-cannot-get-pwgan-to-converge-sound-good]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [May 24, 2020, 9:12am]<br/> <br/> For future reference for anyone struggling; I found that not using v3 of<br/> the PWG discriminator does the trick. I have been able to get it to<br/> sound much better in under one day, using v1<br/> The config I use is this.<br/> <br/> ###########################################################<br/> # FEATURE EXTRACTION SETTING #<br/> ###########################################################<br/> format: 'hdf5'<br/> audio:<br/> clip_norm: true<br/> do_trim_silence: true<br/> frame_length_ms: 50<br/> frame_shift_ms: 12.5<br/> max_norm: 4<br/> hop_length: 275<br/> win_length: 1100<br/> mel_fmax: 8000.0<br/> mel_fmin: 0.0<br/> min_level_db: -100<br/> num_freq: 1025<br/> num_mels: 80<br/> preemphasis: 0.98<br/> ref_level_db: 20<br/> sample_rate: 22050<br/> signal_norm: true<br/> sound_norm: false<br/> symmetric_norm: true<br/> trim_db: 60<br/> <br/> ###########################################################<br/> # GENERATOR NETWORK ARCHITECTURE SETTING #<br/> ###########################################################<br/> generator_params:<br/> in_channels: 1 # Number of input channels.<br/> out_channels: 1 # Number of output channels.<br/> kernel_size: 3 # Kernel size of dilated convolution.<br/> layers: 30 # Number of residual block layers.<br/> stacks: 3 # Number of stacks i.e., dilation cycles.<br/> residual_channels: 64 # Number of channels in residual conv.<br/> gate_channels: 128 # Number of channels in gated conv.<br/> skip_channels: 64 # Number of channels in skip conv.<br/> aux_channels: 80 # Number of channels for auxiliary feature conv.<br/> # Must be the same as num_mels.<br/> aux_context_window: 2 # Context window size for auxiliary feature.<br/> # If set to 2, previous 2 and future 2 frames will be considered.<br/> dropout: 0.0 # Dropout rate. 0.0 means no dropout applied.<br/> use_weight_norm: true # Whether to use weight norm.<br/> # If set to true, it will be applied to all of the conv layers.<br/> upsample_net: 'ConvInUpsampleNetwork' # Upsampling network architecture.<br/> upsample_params: # Upsampling network parameters.<br/> upsample_scales:<br/> - 5<br/> - 5<br/> - 11<br/> <br/> ###########################################################<br/> # DISCRIMINATOR NETWORK ARCHITECTURE SETTING #<br/> ###########################################################<br/> discriminator_params:<br/> in_channels: 1 # Number of input channels.<br/> out_channels: 1 # Number of output channels.<br/> kernel_size: 3 # Number of output channels.<br/> layers: 10 # Number of conv layers.<br/> conv_channels: 64 # Number of chnn layers.<br/> bias: true # Whether to use bias parameter in conv.<br/> use_weight_norm: true # Whether to use weight norm.<br/> # If set to true, it will be applied to all of the conv layers.<br/> nonlinear_activation: 'LeakyReLU' # Nonlinear function after each conv.<br/> nonlinear_activation_params: # Nonlinear function parameters<br/> negative_slope: 0.2 # Alpha in LeakyReLU.<br/> <br/> ###########################################################<br/> # STFT LOSS SETTING #<br/> ###########################################################<br/> stft_loss_params:<br/> fft_sizes: [1024, 2048, 512] # List of FFT size for STFT-based loss.<br/> hop_sizes: [120, 240, 50] # List of hop size for STFT-based loss<br/> win_lengths: [600, 1200, 240] # List of window length for STFT-based loss.<br/> window: 'hann_window' # Window function for STFT-based loss<br/> <br/> ###########################################################<br/> # ADVERSARIAL LOSS SETTING #<br/> ###########################################################<br/> lambda_adv: 4.0 # Loss balancing coefficient.<br/> <br/> ###########################################################<br/> # DATA LOADER SETTING #<br/> ###########################################################<br/> batch_size: 1 # Batch size.<br/> batch_max_steps: 28875 # Length of each audio in batch. Make sure dividable by hop_size.<br/> pin_memory: true # Whether to pin memory in Pytorch DataLoader.<br/> num_workers: 2 # Number of workers in Pytorch DataLoader.<br/> remove_short_samples: true # Whether to remove samples the length of which are less than batch_max_steps.<br/> allow_cache: true # Whether to allow cache in dataset. If true, it requires cpu memory.<br/> <br/> ###########################################################<br/> # OPTIMIZER & SCHEDULER SETTING #<br/> ###########################################################<br/> generator_optimizer_params:<br/> lr: 0.0001 # Generator's learning rate.<br/> eps: 1.0e-6 # Generator's epsilon.<br/> weight_decay: 0.0 # Generator's weight decay coefficient.<br/> generator_scheduler_params:<br/> step_size: 200000 # Generator's scheduler step size.<br/> gamma: 0.5 # Generator's scheduler gamma.<br/> # At each step size, lr will be multiplied by this parameter.<br/> generator_grad_norm: 10 # Generator's gradient norm.<br/> discriminator_optimizer_params:<br/> lr: 0.00005 # Discriminator's learning rate.<br/> eps: 1.0e-6 # Discriminator's epsilon.<br/> weight_decay: 0.0 # Discriminator's weight decay coefficient.<br/> discriminator_scheduler_params:<br/> step_size: 200000 # Discriminator's scheduler step size.<br/> gamma: 0.5 # Discriminator's scheduler gamma.<br/> # At each step size, lr will be multiplied by this parameter.<br/> discriminator_grad_norm: 1 # Discriminator's gradient norm.<br/> <br/> ###########################################################<br/> # INTERVAL SETTING #<br/> ###########################################################<br/> discriminator_train_start_steps: 100000 # Number of steps to start to train discriminator.<br/> train_max_steps: 400000 # Number of training steps.<br/> save_interval_steps: 5000 # Interval steps to save checkpoint.<br/> eval_interval_steps: 1000 # Interval steps to evaluate the network.<br/> log_interval_steps: 100 # Interval steps to record the training log.<br/> <br/> ###########################################################<br/> # OTHER SETTING #<br/> ###########################################################<br/> num_save_intermediate_results: 4 # Number of results to be saved as intermediate results.<br/> <br/> It is a mix of an original config and<br/> 's tts configs found on the<br/> fork.<br/> <br/> I am training on a private speaker and will let it train until it<br/> reaches the full 400k mark and then decide if I go with 1000k. My goal<br/> is that, if this works, I want to train a LibriTTS universal vocoder on<br/> 22050Hz and release it here. We will see how it goes<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> othiele<br/> <DATE> [May 24, 2020, 9:23am]<br/> <br/> Thanks, that helps a lot. Would be great to have a vocoder to try.<br/> erogols old WaveRNN one with 16 KHz shows what it can do, just sounds<br/> like Micky<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> hamza.mughal<br/> <DATE> [October 12, 2020, 7:44am]<br/> <br/> can you please point<br/> me to the specific commit in mozilla/TTS repo for v1 of PWG<br/> discriminator. *slash* <br/> I am not able to locate what you referred here.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [October 12, 2020, 11:28am]<br/> <br/> Hi, I got it from the original repo, but it is a hassle to edit it to<br/> fit the mozilla fork, so I would instead train using the integrated<br/> vocoder module.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> hamza.mughal<br/> <DATE> [October 12, 2020, 5:25pm]<br/> <br/> I confirmed the<br/> parameters of parallel_wavegan_discriminator I have used the same ones<br/> for training. *slash* <br/> Thanks alot!<br/> <br/> [Archived Post]<br/>
